### PR TITLE
refactor to jobs

### LINKED ISSUE #926: Change logging/naming of "transactions" (internal node TXs vs Ethereum TXs)
It is really tough to know which "transaction" is being talked about in logs and API returns.

* Change the logging to be clear when the log entry talks about an "ethereum transaction" (instead of a node internal TX).
* Change the name of the "nternal TX ("transaction") to "taskId" (or something else clear)-> making sure to distinguish the async tasks compared to transactions sent to a blockchain/Ethereum

### PR DESCRIPTION
**Pull request by vedhavyas**
closes #926

### CODE REVIEW

File: anchors/service.go [hunk 1]
```diff
@@ -94,7 +94,7 @@ func (s *service) PreCommitAnchor(ctx context.Context, anchorID AnchorID, signin
 	}
 
 	log.Infof("Add Anchor to Pre-commit %s from did:%s", anchorID.String(), did.ToAddress().String())
-	_, done, err := s.txManager.ExecuteWithinTX(ctx, did, txID, "Check TX for anchor commit",
+	_, done, err := s.txManager.ExecuteWithinJob(ctx, did, txID, "Check TX for anchor commit",
```
Comment (vimukthi-git) [inline anchors/service.go]: rename variable

File: anchors/service.go [hunk 2]
```diff
@@ -104,8 +104,8 @@ func (s *service) PreCommitAnchor(ctx context.Context, anchorID AnchorID, signin
 }
 
 // ethereumTX is submitting an Ethereum transaction and starts a task to wait for the transaction result
-func (s service) ethereumTX(opts *bind.TransactOpts, contractMethod interface{}, params ...interface{}) func(accountID identity.DID, txID transactions.TxID, txMan transactions.Manager, errOut chan<- error) {
-	return func(accountID identity.DID, txID transactions.TxID, txMan transactions.Manager, errOut chan<- error) {
+func (s service) ethereumTX(opts *bind.TransactOpts, contractMethod interface{}, params ...interface{}) func(accountID identity.DID, txID jobs.JobID, txMan jobs.Manager, errOut chan<- error) {
```
Comment (vimukthi-git) [inline anchors/service.go]: rename variable

File: anchors/service.go [hunk 3]
```diff
@@ -32,10 +32,10 @@ type service struct {
 	anchorRepositoryContract anchorRepositoryContract
 	client                   ethereum.Client
 	queue                    *queue.Server
-	txManager                transactions.Manager
+	txManager                jobs.Manager
```
Comment (vimukthi-git) [inline anchors/service.go]: rename variable

File: contextutil/context.go [hunk 4]
```diff
@@ -30,15 +30,15 @@ func New(ctx context.Context, cfg config.Account) (context.Context, error) {
...
 // WithTX returns a context with TX ID
-func WithTX(ctx context.Context, txID transactions.TxID) context.Context {
+func WithTX(ctx context.Context, txID jobs.JobID) context.Context {
 	return context.WithValue(ctx, tx, txID)
 }
 
 // TX returns current txID
-func TX(ctx context.Context) transactions.TxID {
-	tid, ok := ctx.Value(tx).(transactions.TxID)
+func TX(ctx context.Context) jobs.JobID {
```
Comment (vimukthi-git) [inline contextutil/context.go]: rename func

File: documents/anchor_task.go [hunk 5]
```diff
@@ -121,14 +121,14 @@ func (d *documentAnchorTask) RunTask() (res interface{}, err error) {
 }
 
 // InitDocumentAnchorTask enqueues a new document anchor task for a given combination of accountID/modelID/txID.
-func InitDocumentAnchorTask(txMan transactions.Manager, tq queue.TaskQueuer, accountID identity.DID, modelID []byte, txID transactions.TxID) (queue.TaskResult, error) {
+func InitDocumentAnchorTask(txMan jobs.Manager, tq queue.TaskQueuer, accountID identity.DID, modelID []byte, txID jobs.JobID) (queue.TaskResult, error) {
```
Comment (vimukthi-git) [inline documents/anchor_task.go]: rename variable

File: documents/anchor_task.go [hunk 6]
```diff
@@ -142,8 +142,8 @@ func InitDocumentAnchorTask(txMan transactions.Manager, tq queue.TaskQueuer, acc
 }
 
 // CreateAnchorTransaction creates a transaction for anchoring a document using transaction manager
-func CreateAnchorTransaction(txMan transactions.Manager, tq queue.TaskQueuer, self identity.DID, txID transactions.TxID, documentID []byte) (transactions.TxID, chan bool, error) {
-	txID, done, err := txMan.ExecuteWithinTX(context.Background(), self, txID, "anchor document", func(accountID identity.DID, TID transactions.TxID, txMan transactions.Manager, errChan chan<- error) {
+func CreateAnchorTransaction(txMan jobs.Manager, tq queue.TaskQueuer, self identity.DID, txID jobs.JobID, documentID []byte) (jobs.JobID, chan bool, error) {
```
Comment (vimukthi-git) [inline documents/anchor_task.go]: rename variable

File: documents/entity/bootstrapper.go [hunk 7]
```diff
@@ -44,7 +44,7 @@ func (Bootstrapper) Bootstrap(ctx map[string]interface{}) error {
 		return errors.New("queue server not initialised")
 	}
 
-	txManager, ok := ctx[transactions.BootstrappedService].(transactions.Manager)
+	txManager, ok := ctx[jobs.BootstrappedService].(jobs.Manager)
```
Comment (vimukthi-git) [inline documents/entity/bootstrapper.go]: rename variable

File: jobs/job.go [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+// Config is the config interface for jobs package
+type Config interface {
+	GetEthereumContextWaitTimeout() time.Duration
+}
+
+// JobManager is a manager for centrifuge Jobs.
+type Manager interface {
+	// ExecuteWithinJob executes the given unit of work within a Job
+	ExecuteWithinJob(ctx context.Context, accountID identity.DID, existingJobID JobID, desc string, work func(accountID identity.DID, jobID JobID, txMan Manager, err chan<- error)) (jobID JobID, done chan bool, err error)
```
Comment (mikiquantum) [inline jobs/job.go]:
```suggestion
	ExecuteWithinJob(ctx context.Context, accountID identity.DID, existingJobID JobID, desc string, work func(accountID identity.DID, jobID JobID, jobMan Manager, err chan<- error)) (jobID JobID, done chan bool, err error)
```

File: documents/entity/model_test.go [hunk 9]
```diff
@@ -66,10 +66,10 @@ func TestMain(m *testing.M) {
 	ethClient := &testingcommons.MockEthClient{}
 	ethClient.On("GetEthClient").Return(nil)
 	ctx[ethereum.BootstrappedEthereumClient] = ethClient
-	txMan := &testingtx.MockTxManager{}
-	ctx[transactions.BootstrappedService] = txMan
+	txMan := &testingjobs.MockJobManager{}
```
Comment (vimukthi-git) [inline documents/entity/model_test.go]: rename variable

File: documents/entity/service.go [hunk 10]
```diff
@@ -74,7 +74,7 @@ func DefaultService(
 	srv documents.Service,
 	repo documents.Repository,
 	queueSrv queue.TaskQueuer,
-	txManager transactions.Manager,
+	txManager jobs.Manager,
```
Comment (vimukthi-git) [inline documents/entity/service.go]: rename variable

File: documents/entityrelationship/bootstrapper.go [hunk 11]
```diff
@@ -49,7 +49,7 @@ func (Bootstrapper) Bootstrap(ctx map[string]interface{}) error {
 		return errors.New("queue server not initialised")
 	}
 
-	txManager, ok := ctx[transactions.BootstrappedService].(transactions.Manager)
+	txManager, ok := ctx[jobs.BootstrappedService].(jobs.Manager)
```
Comment (vimukthi-git) [inline documents/entityrelationship/bootstrapper.go]: rename variable

File: documents/entityrelationship/model_test.go [hunk 12]
```diff
@@ -49,10 +49,10 @@ func TestMain(m *testing.M) {
 	ethClient := &testingcommons.MockEthClient{}
 	ethClient.On("GetEthClient").Return(nil)
 	ctx[ethereum.BootstrappedEthereumClient] = ethClient
-	txMan := &testingtx.MockTxManager{}
-	ctx[transactions.BootstrappedService] = txMan
+	txMan := &testingjobs.MockJobManager{}
```
Comment (vimukthi-git) [inline documents/entityrelationship/model_test.go]: rename variable

File: jobs/jobsv1/base_task.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+	// state
+	JobManager jobs.Manager
+}
+
+// ParseJobID parses txID.
+func (b *BaseTask) ParseJobID(taskTypeName string, kwargs map[string]interface{}) error {
+	txID, ok := kwargs[jobs.JobIDParam].(string)
+	if !ok {
+		return errors.New("missing transaction ID")
```
Comment (mikiquantum) [inline jobs/jobsv1/base_task.go]:
```suggestion
		return errors.New("missing job ID")
```

File: documents/entityrelationship/service.go [hunk 14]
```diff
@@ -41,7 +41,7 @@ type service struct {
 	documents.Service
 	repo      repository
 	queueSrv  queue.TaskQueuer
-	txManager transactions.Manager
+	txManager jobs.Manager
```
Comment (vimukthi-git) [inline documents/entityrelationship/service.go]: rename variable

File: documents/entityrelationship/service.go [hunk 15]
```diff
@@ -50,7 +50,7 @@ func DefaultService(
 	srv documents.Service,
 	repo repository,
 	queueSrv queue.TaskQueuer,
-	txManager transactions.Manager,
+	txManager jobs.Manager,
```
Comment (vimukthi-git) [inline documents/entityrelationship/service.go]: rename variable

File: jobs/jobsv1/base_task.go [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+// BaseTask holds the required details and helper functions for tasks to update jobs.
+// should be embedded into the task
+type BaseTask struct {
+	JobID jobs.JobID
+
+	// state
+	JobManager jobs.Manager
+}
+
+// ParseJobID parses txID.
```
Comment (mikiquantum) [inline jobs/jobsv1/base_task.go]:
```suggestion
// ParseJobID parses jobID.
```

File: jobs/jobsv1/base_task.go [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+func (b *BaseTask) ParseJobID(taskTypeName string, kwargs map[string]interface{}) error {
+	txID, ok := kwargs[jobs.JobIDParam].(string)
+	if !ok {
+		return errors.New("missing transaction ID")
+	}
+
+	var err error
+	b.JobID, err = jobs.FromString(txID)
+	if err != nil {
+		return errors.New("invalid transaction ID")
```
Comment (mikiquantum) [inline jobs/jobsv1/base_task.go]:
```suggestion
		return errors.New("invalid job ID")
```

File: documents/invoice/bootstrapper.go [hunk 18]
```diff
@@ -41,7 +41,7 @@ func (Bootstrapper) Bootstrap(ctx map[string]interface{}) error {
 		return errors.New("queue server not initialised")
 	}
 
-	txManager, ok := ctx[transactions.BootstrappedService].(transactions.Manager)
+	txManager, ok := ctx[jobs.BootstrappedService].(jobs.Manager)
```
Comment (vimukthi-git) [inline documents/invoice/bootstrapper.go]: rename variable

File: documents/invoice/service.go [hunk 19]
```diff
@@ -36,7 +36,7 @@ type service struct {
 	documents.Service
 	repo           documents.Repository
 	queueSrv       queue.TaskQueuer
-	txManager      transactions.Manager
+	txManager      jobs.Manager
```
Comment (vimukthi-git) [inline documents/invoice/service.go]: rename variable

File: jobs/jobsv1/base_task.go [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+type BaseTask struct {
+	JobID jobs.JobID
+
+	// state
+	JobManager jobs.Manager
+}
+
+// ParseJobID parses txID.
+func (b *BaseTask) ParseJobID(taskTypeName string, kwargs map[string]interface{}) error {
+	txID, ok := kwargs[jobs.JobIDParam].(string)
```
Comment (mikiquantum) [inline jobs/jobsv1/base_task.go]:
```suggestion
	jobID, ok := kwargs[jobs.JobIDParam].(string)
```

File: documents/purchaseorder/bootstrapper.go [hunk 21]
```diff
@@ -41,7 +41,7 @@ func (Bootstrapper) Bootstrap(ctx map[string]interface{}) error {
 		return errors.New("queue server not initialised")
 	}
 
-	txManager, ok := ctx[transactions.BootstrappedService].(transactions.Manager)
+	txManager, ok := ctx[jobs.BootstrappedService].(jobs.Manager)
```
Comment (vimukthi-git) [inline documents/purchaseorder/bootstrapper.go]: rename variable

File: documents/purchaseorder/model_test.go [hunk 22]
```diff
@@ -48,10 +48,10 @@ func TestMain(m *testing.M) {
 	ethClient := &testingcommons.MockEthClient{}
 	ethClient.On("GetEthClient").Return(nil)
 	ctx[ethereum.BootstrappedEthereumClient] = ethClient
-	txMan := &testingtx.MockTxManager{}
-	ctx[transactions.BootstrappedService] = txMan
+	txMan := &testingjobs.MockJobManager{}
```
Comment (vimukthi-git) [inline documents/purchaseorder/model_test.go]: rename variable

File: documents/purchaseorder/service.go [hunk 23]
```diff
@@ -36,7 +36,7 @@ type service struct {
 	documents.Service
 	repo           documents.Repository
 	queueSrv       queue.TaskQueuer
-	txManager      transactions.Manager
+	txManager      jobs.Manager
```
Comment (vimukthi-git) [inline documents/purchaseorder/service.go]: rename variable

File: jobs/jobsv1/base_task.go [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+	if err != nil {
+		return errors.New("invalid transaction ID")
+	}
+
+	log.Infof("Task %s parsed for tx: %s\n", taskTypeName, b.JobID)
+	return nil
+}
+
+// UpdateTransaction add a new log and updates the status of the job based on the error.
+func (b *BaseTask) UpdateTransaction(accountID identity.DID, taskTypeName string, err error) error {
```
Comment (mikiquantum) [inline jobs/jobsv1/base_task.go]:
```suggestion
func (b *BaseTask) UpdateJob(accountID identity.DID, taskTypeName string, err error) error {
```

File: documents/purchaseorder/service_test.go [hunk 25]
```diff
@@ -49,7 +49,7 @@ func getServiceWithMockedLayers() (*testingcommons.MockIdentityService, Service)
 	idService.On("IsSignedWithPurpose", mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(true, nil).Once()
 	queueSrv := new(testingutils.MockQueue)
 	queueSrv.On("EnqueueJob", mock.Anything, mock.Anything).Return(&gocelery.AsyncResult{}, nil)
-	txManager := ctx[transactions.BootstrappedService].(transactions.Manager)
+	txManager := ctx[jobs.BootstrappedService].(jobs.Manager)
```
Comment (vimukthi-git) [inline documents/purchaseorder/service_test.go]: rename variable

File: ethereum/geth_client_integration_test.go [hunk 26]
```diff
@@ -31,7 +31,7 @@ var ctx = map[string]interface{}{}
 
 func registerMockedTransactionTask() {
 	queueSrv := ctx[bootstrap.BootstrappedQueueServer].(*queue.Server)
-	txManager := ctx[transactions.BootstrappedService].(transactions.Manager)
+	txManager := ctx[jobs.BootstrappedService].(jobs.Manager)
```
Comment (vimukthi-git) [inline ethereum/geth_client_integration_test.go]: rename variable

File: jobs/jobsv1/base_task.go [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+	return b.UpdateTransactionWithValue(accountID, taskTypeName, err, nil)
+}
+
+// UpdateJobWithValue add a new log and updates the status of the transaction based on the error and adds a value to the tx
+func (b *BaseTask) UpdateTransactionWithValue(accountID identity.DID, taskTypeName string, err error, txValue *jobs.JobValue) error {
```
Comment (mikiquantum) [inline jobs/jobsv1/base_task.go]:
```suggestion
func (b *BaseTask) UpdateJobWithValue(accountID identity.DID, taskTypeName string, err error, jobValue *jobs.JobValue) error {
```

File: ethereum/transaction_status_task_integration_test.go [hunk 28]
```diff
@@ -6,24 +6,23 @@ import (
...
-	"github.com/centrifuge/go-centrifuge/transactions"
+	"github.com/centrifuge/go-centrifuge/testingutils/identity"
 	"github.com/stretchr/testify/assert"
 )
 
-func enqueueJob(t *testing.T, txHash string) (transactions.Manager, identity.DID, transactions.TxID, chan bool) {
+func enqueueJob(t *testing.T, txHash string) (jobs.Manager, identity.DID, jobs.JobID, chan bool) {
 	queueSrv := ctx[bootstrap.BootstrappedQueueServer].(*queue.Server)
-	txManager := ctx[transactions.BootstrappedService].(transactions.Manager)
+	txManager := ctx[jobs.BootstrappedService].(jobs.Manager)
```
Comment (vimukthi-git) [inline ethereum/transaction_status_task_integration_test.go]: rename variable

File: identity/ideth/bootstrapper.go [hunk 29]
```diff
@@ -38,7 +35,7 @@ func (*Bootstrapper) Bootstrap(context map[string]interface{}) error {
 		return err
 	}
 
-	txManager, ok := context[transactions.BootstrappedService].(transactions.Manager)
+	txManager, ok := context[jobs.BootstrappedService].(jobs.Manager)
```
Comment (vimukthi-git) [inline identity/ideth/bootstrapper.go]: rename variable

File: identity/ideth/factory.go [hunk 30]
```diff
@@ -24,13 +24,13 @@ type factory struct {
 	factoryAddress  common.Address
 	factoryContract *FactoryContract
 	client          ethereum.Client
-	txManager       transactions.Manager
+	txManager       jobs.Manager
```
Comment (vimukthi-git) [inline identity/ideth/factory.go]: rename variable

File: identity/ideth/service.go [hunk 31]
```diff
@@ -59,7 +59,7 @@ func methodToOp(method string) config.ContractOp {
 
 type service struct {
 	client    ethereum.Client
-	txManager transactions.Manager
+	txManager jobs.Manager
```
Comment (vimukthi-git) [inline identity/ideth/service.go]: rename variable

File: jobs/jobsv1/handler.go [hunk 32]
```diff
@@ -7,41 +7,41 @@ import (
...
+// ErrInvalidJobID error for Invalid transaction ID.
+const ErrInvalidJobID = errors.Error("Invalid Job ID")
 
 // ErrInvalidAccountID error for Invalid account ID.
 const ErrInvalidAccountID = errors.Error("Invalid Tenant ID")
 
-var apiLog = logging.Logger("transaction-api")
+var apiLog = logging.Logger("jobs-api")
 
 // GRPCHandler returns an implementation of the TransactionServiceServer
```
Comment (mikiquantum) [inline jobs/jobsv1/handler.go]:
```suggestion
// GRPCHandler returns an implementation of the JobServiceServer
```

File: identity/ideth/service.go [hunk 33]
```diff
@@ -108,7 +108,7 @@ func (i service) bindContract(did id.DID) (contract, error) {
 }
 
 // NewService creates a instance of the identity service
-func NewService(client ethereum.Client, txManager transactions.Manager, queue *queue.Server, conf id.Config) id.ServiceDID {
+func NewService(client ethereum.Client, txManager jobs.Manager, queue *queue.Server, conf id.Config) id.ServiceDID {
```
Comment (vimukthi-git) [inline identity/ideth/service.go]: rename variable

File: identity/ideth/service_integration_test.go [hunk 34]
```diff
@@ -32,7 +28,7 @@ func getTestKey() id.KeyDID {
 
 func initIdentity() id.ServiceDID {
 	client := ctx[ethereum.BootstrappedEthereumClient].(ethereum.Client)
-	txManager := ctx[transactions.BootstrappedService].(transactions.Manager)
+	txManager := ctx[jobs.BootstrappedService].(jobs.Manager)
```
Comment (vimukthi-git) [inline identity/ideth/service_integration_test.go]: rename variable

File: jobs/jobsv1/handler.go [hunk 35]
```diff
@@ -7,41 +7,41 @@ import (
...
-func GRPCHandler(srv transactions.Manager, configService config.Service) transactionspb.TransactionServiceServer {
+func GRPCHandler(srv jobs.Manager, configService config.Service) jobspb.JobServiceServer {
 	return grpcHandler{srv: srv, configService: configService}
 }
 
 // grpcHandler implements transactionspb.TransactionServiceServer
```
Comment (mikiquantum) [inline jobs/jobsv1/handler.go]:
```suggestion
// grpcHandler implements jobspb.JobServiceServer
```

File: jobs/jobsv1/handler.go [hunk 36]
```diff
@@ -7,41 +7,41 @@ import (
...
 type grpcHandler struct {
-	srv           transactions.Manager
+	srv           jobs.Manager
 	configService config.Service
 }
 
-// GetTransactionStatus returns transaction status of the given transaction id.
-func (h grpcHandler) GetTransactionStatus(ctx context.Context, req *transactionspb.TransactionStatusRequest) (*transactionspb.TransactionStatusResponse, error) {
+// GetJobStatus returns transaction status of the given transaction id.
```
Comment (mikiquantum) [inline jobs/jobsv1/handler.go]:
```suggestion
// GetJobStatus returns job status of the given job id.
```

File: jobs/job.go (hunk 8 above)
Comment (vimukthi-git) [inline jobs/job.go]: rename variable

File: jobs/jobsv1/handler_test.go [hunk 37]
```diff
@@ -0,0 +1,56 @@
...
+	"github.com/centrifuge/go-centrifuge/config"
+	"github.com/centrifuge/go-centrifuge/errors"
+	"github.com/centrifuge/go-centrifuge/identity"
+	"github.com/centrifuge/go-centrifuge/jobs"
+	"github.com/centrifuge/go-centrifuge/protobufs/gen/go/jobs"
+	"github.com/centrifuge/go-centrifuge/testingutils/config"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGRPCHandler_GetTransactionStatus(t *testing.T) {
```
Comment (mikiquantum) [inline jobs/jobsv1/handler_test.go]:
```suggestion
func TestGRPCHandler_GetJobStatus(t *testing.T) {
```

File: jobs/jobsv1/manager.go [hunk 38]
```diff
@@ -0,0 +1,192 @@
...
+	createJob(accountID identity.DID, desc string) (*jobs.Job, error)
+}
+
+// NewManager returns a JobManager implementation.
+func NewManager(config jobs.Config, repo jobs.Repository) jobs.Manager {
+	return &manager{config: config, repo: repo}
+}
+
+// manager implements JobManager.
+// TODO [TXManager] convert this into an implementation of node.Server and start it at node start so that we can bring down transaction go routines cleanly
```
Comment (vimukthi-git) [inline jobs/jobsv1/manager.go]: fix comment

File: jobs/jobsv1/manager.go [hunk 39]
```diff
@@ -0,0 +1,192 @@
...
+	}
+
+	// status particular to the task
+	tx.TaskStatus[taskName] = status
+	tx.Logs = append(tx.Logs, jobs.NewLog(taskName, message))
+	return s.saveJob(tx)
+}
+
+// ExecuteWithinJob executes a task within a Job.
+func (s *manager) ExecuteWithinJob(ctx context.Context, accountID identity.DID, existingTxID jobs.JobID, desc string, work func(accountID identity.DID, txID jobs.JobID, txMan jobs.Manager, err chan<- error)) (txID jobs.JobID, done chan bool, err error) {
```
Comment (vimukthi-git) [inline jobs/jobsv1/manager.go]: rename variable

File: jobs/jobsv1/manager.go [hunk 40]
```diff
@@ -0,0 +1,192 @@
...
+const (
+	managerLogPrefix = "manager"
+)
+
+// extendedManager exposes package specific functions.
+type extendedManager interface {
+	jobs.Manager
+
+	// saveJob only exposed for testing within package.
+	// DO NOT use this outside of the package, use ExecuteWithinJob to initiate a transaction with management.
```
Comment (mikiquantum) [inline jobs/jobsv1/manager.go]:
```suggestion
	// DO NOT use this outside of the package, use ExecuteWithinJob to initiate a job with management.
```

File: jobs/jobsv1/manager_test.go [hunk 41]
```diff
@@ -25,124 +24,124 @@ func (mockConfig) GetEthereumContextWaitTimeout() time.Duration {
 
 func TestService_ExecuteWithinTX_happy(t *testing.T) {
 	cid := testingidentity.GenerateRandomDID()
-	srv := ctx[transactions.BootstrappedService].(transactions.Manager)
-	tid, done, err := srv.ExecuteWithinTX(context.Background(), cid, transactions.NilTxID(), "", func(accountID identity.DID, txID transactions.TxID, txMan transactions.Manager, err chan<- error) {
+	srv := ctx[jobs.BootstrappedService].(jobs.Manager)
+	tid, done, err := srv.ExecuteWithinJob(context.Background(), cid, jobs.NilJobID(), "", func(accountID identity.DID, txID jobs.JobID, txMan jobs.Manager, err chan<- error) {
```
Comment (vimukthi-git) [inline jobs/jobsv1/manager_test.go]: rename variable

File: jobs/jobsv1/manager_test.go [hunk 42]
```diff
@@ -25,124 +24,124 @@ func (mockConfig) GetEthereumContextWaitTimeout() time.Duration {
...
+	assert.Equal(t, jobs.Success, trn.Status)
 }
 
 func TestService_ExecuteWithinTX_err(t *testing.T) {
 	errStr := "dummy"
 	cid := testingidentity.GenerateRandomDID()
-	srv := ctx[transactions.BootstrappedService].(transactions.Manager)
-	tid, done, err := srv.ExecuteWithinTX(context.Background(), cid, transactions.NilTxID(), "SomeTask", func(accountID identity.DID, txID transactions.TxID, txMan transactions.Manager, err chan<- error) {
+	srv := ctx[jobs.BootstrappedService].(jobs.Manager)
+	tid, done, err := srv.ExecuteWithinJob(context.Background(), cid, jobs.NilJobID(), "SomeTask", func(accountID identity.DID, txID jobs.JobID, txMan jobs.Manager, err chan<- error) {
```
Comment (vimukthi-git) [inline jobs/jobsv1/manager_test.go]: rename variable

File: nft/bootstrapper.go [hunk 43]
```diff
@@ -42,7 +42,7 @@ func (*Bootstrapper) Bootstrap(ctx map[string]interface{}) error {
 	}
 	queueSrv := ctx[bootstrap.BootstrappedQueueServer].(*queue.Server)
 
-	txManager, ok := ctx[transactions.BootstrappedService].(transactions.Manager)
+	txManager, ok := ctx[jobs.BootstrappedService].(jobs.Manager)
```
Comment (vimukthi-git) [inline nft/bootstrapper.go]: rename variable

File: jobs/jobsv1/manager.go (hunk 39 above)
Comment (mikiquantum) [inline jobs/jobsv1/manager.go]:
```suggestion
func (s *manager) ExecuteWithinJob(ctx context.Context, accountID identity.DID, existingTxID jobs.JobID, desc string, work func(accountID identity.DID, jobID jobs.JobID, jobMan jobs.Manager, err chan<- error)) (txID jobs.JobID, done chan bool, err error) {
```

File: nft/eth_invoice_unpaid.go [hunk 44]
```diff
@@ -196,20 +196,20 @@ func (s *ethInvoiceUnpaid) MintNFT(ctx context.Context, req MintNFTRequest) (*Mi
...
-		TransactionID: txID.String(),
-		TokenID:       tokenID.String(),
+		JobID:   txID.String(),
+		TokenID: tokenID.String(),
 	}, done, nil
 }
 
-func (s *ethInvoiceUnpaid) minter(ctx context.Context, tokenID TokenID, model documents.Model, req MintNFTRequest) func(accountID identity.DID, txID transactions.TxID, txMan transactions.Manager, errOut chan<- error) {
-	return func(accountID identity.DID, txID transactions.TxID, txMan transactions.Manager, errOut chan<- error) {
+func (s *ethInvoiceUnpaid) minter(ctx context.Context, tokenID TokenID, model documents.Model, req MintNFTRequest) func(accountID identity.DID, txID jobs.JobID, txMan jobs.Manager, errOut chan<- error) {
```
Comment (vimukthi-git) [inline nft/eth_invoice_unpaid.go]: rename variable

File: nft/invoice_unpaid_integration_test.go [hunk 45]
```diff
@@ -36,7 +36,7 @@ var cfgService config.Service
 var idService identity.ServiceDID
 var idFactory identity.Factory
 var InvoiceUnpaid nft.InvoiceUnpaid
-var txManager transactions.Manager
+var txManager jobs.Manager
```
Comment (vimukthi-git) [inline nft/invoice_unpaid_integration_test.go]: rename variable

File: testingutils/testingjobs/mockjob.go [hunk 46]
```diff
@@ -0,0 +1,46 @@
...
+
+func (m MockJobManager) UpdateJobWithValue(accountID identity.DID, id jobs.JobID, key string, value []byte) error {
+	panic("implement me")
+}
+
+func (m MockJobManager) UpdateTaskStatus(accountID identity.DID, id jobs.JobID, status jobs.Status, taskName, message string) error {
+	panic("implement me")
+}
+
+func (m MockJobManager) ExecuteWithinJob(ctx context.Context, accountID identity.DID, existingTxID jobs.JobID, desc string, work func(accountID identity.DID, txID jobs.JobID, txMan jobs.Manager, err chan<- error)) (txID jobs.JobID, done chan bool, err error) {
```
Comment (vimukthi-git) [inline testingutils/testingjobs/mockjob.go]: rename variable

File: jobs/jobsv1/manager.go [hunk 47]
```diff
@@ -0,0 +1,192 @@
...
+			if e != nil {
+				log.Error(e)
+			}
+		}
+		done <- true
+	}(ctx)
+	return t.ID, done, nil
+}
+
+// saveJob saves the transaction.
```
Comment (mikiquantum) [inline jobs/jobsv1/manager.go]:
```suggestion
// saveJob saves the job.
```

File: jobs/jobsv1/manager.go [hunk 48]
```diff
@@ -0,0 +1,192 @@
...
+func (s *manager) saveJob(tx *jobs.Job) error {
+	err := s.repo.Save(tx)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetJob returns the transaction associated with identity and id.
```
Comment (mikiquantum) [inline jobs/jobsv1/manager.go]:
```suggestion
// GetJob returns the job associated with identity and id.
```

File: jobs/jobsv1/manager.go [hunk 49]
```diff
@@ -0,0 +1,192 @@
...
+func (s *manager) GetJob(accountID identity.DID, id jobs.JobID) (*jobs.Job, error) {
+	return s.repo.Get(accountID, id)
+}
+
+// createJob creates a new transaction and saves it to the DB.
```
Comment (mikiquantum) [inline jobs/jobsv1/manager.go]:
```suggestion
// createJob creates a new job and saves it to the DB.
```

File: jobs/jobsv1/manager.go [hunk 50]
```diff
@@ -0,0 +1,192 @@
...
+func (s *manager) createJob(accountID identity.DID, desc string) (*jobs.Job, error) {
+	tx := jobs.NewJob(accountID, desc)
+	return tx, s.saveJob(tx)
+}
+
+// WaitForJob blocks until job status is moved from pending state.
+// Note: use it with caution as this will block.
+func (s *manager) WaitForJob(accountID identity.DID, txID jobs.JobID) error {
```
Comment (mikiquantum) [inline jobs/jobsv1/manager.go]:
```suggestion
func (s *manager) WaitForJob(accountID identity.DID, jobID jobs.JobID) error {
```

File: jobs/jobsv1/manager.go [hunk 51]
```diff
@@ -0,0 +1,192 @@
...
+		case jobs.Success:
+			return nil
+		default:
+			time.Sleep(10 * time.Millisecond)
+			continue
+		}
+	}
+}
+
+// GetJobStatus returns the transaction status associated with identity and id.
```
Comment (mikiquantum) [inline jobs/jobsv1/manager.go]:
```suggestion
// GetJobStatus returns the job status associated with identity and id.
```

**Pull request merged**